### ISSUE TITLE
How to calculate MFU as shown in readme?

### ISSUE THREAD

**Issue by amankhandelia** [google/maxtext#8]
Question for the developers, as in the [readme page ](https://github.com/google/maxtext/blob/3d236676955713b5647b9048c0eeac8fec1a804d/README.md#runtime-performance-results), one can see the MFU for the TPU VM.

Say if I were to make any modifications and run the same on a TPU VM, how can I calculate the MFU, in order to benchmark my changes with reference to performance gain or loss?

PS: I did go through the code to find any script calculating the MFU, but I couldn't, if I have missed something please help me find the same.

**Comment by rwitten** [comment]
Great questions. The model flops are calculated in train.py as `6 * parameters * batch in tokens`. Then when we divide model flops by step times we compute `per device TFLOPs / second`. To calculate MFUs, you need to divide the `per device TFLOPs/second` by the hardware peak tflops (275 TFLOP/s on the `v4-8`).

Please let me know if you have further questions.

https://github.com/google/maxtext/blob/main/MaxText/train.py#L79-L83

**Comment by amankhandelia** [comment]
Hey Rafi (@rwitten),

Thanks for the elaborate and helpful response, I have one follow up question

I am running the base.yml on `v3-8` and I am getting the following logs

```
completed step: 15107, seconds: 2.167, TFLOP/s: 36.722, loss: 2.191
completed step: 15108, seconds: 2.164, TFLOP/s: 36.771, loss: 2.173
completed step: 15109, seconds: 2.166, TFLOP/s: 36.739, loss: 2.191
completed step: 15110, seconds: 2.167, TFLOP/s: 36.725, loss: 2.285
completed step: 15111, seconds: 2.167, TFLOP/s: 36.729, loss: 2.155
completed step: 15112, seconds: 2.166, TFLOP/s: 36.741, loss: 2.297
completed step: 15113, seconds: 2.166, TFLOP/s: 36.743, loss: 2.154
completed step: 15114, seconds: 2.169, TFLOP/s: 36.692, loss: 2.179
completed step: 15115, seconds: 2.165, TFLOP/s: 36.762, loss: 2.270
```

So, given the above logs, as you can see `per device TFLOPs / second` is `37`ish, and peak performance of the `v3-8` from the [tpu docs](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu_v3) is 
<html>
<body>


Peak compute per chip | 123 teraflops (bf16)
-- | --



</body>
</html>

So based on the formula suggested above `MFU = (36.7/123) * 100 = 29.83`, 

Hence my question, is this correct calculation, my current MFU is 29.83%?

**Comment by rwitten** [comment]
Great question! Can you double check your `len(jax.devices())`? I believe you will see 8 devices. MaxText reports your TFLOP/s/device.

A v3 chip has ~123 peak TFLOP/s and a v3 core has ~61.5 peak TFLOP/s.

Therefore If you see 8 devices, that means MaxText is reporting your TFLOP/s per CORE, since a `v3-8` has 4 chips and 8 cores. If MaxText is reporting TFLOP/s per CORE, your MFU is twice what you're calculating. If MaxText is reporting TFLOP/s per CHIP, your MFU calculation is right.

Please let me know what you find. I haven't had the chance to run this on v3.

**Comment by amankhandelia** [comment]
Yes you are right, it has 8 devices
```
>>> import jax
>>> len(jax.devices())
8
```

I guess than, the MFU calculation changes to `MFU = (36.7/61.5) * 100 = 59.66`

By the way, thanks for all the quick and detailed responses.

**Comment by rwitten** [comment]
Glad you're all set here! Feel free to reach out with questions any time.

(For folks coming across this after the fact, Aman achieved an MFU of `59.66%` on a `v3-8`.)
